### PR TITLE
card_virtuals table

### PR DESCRIPTION
**Pull request by xithan**

### CODE REVIEW

File: card/mod/admin/set/self/admin.rb [hunk 1]
```diff
@@ -30,7 +30,7 @@ def run_task_from_task_basket task
   abort :success
 end
 
-def not_alRAlowed task
+def not_allowed task
```
Comment (ethn) [inline card/mod/admin/set/self/admin.rb]: Won't this break all of our `not_alRAlowed` calls?  ;)

File: card/mod/admin/set/self/admin.rb [hunk 2]
```diff
@@ -118,7 +118,7 @@ def solid_cache_count
   end
 
   def machine_cache_count
-    Card.search right: { codename: "machine_cache" }, return: "count"
+    Card::Virtual.where(right_id: Card.fetch_id(:machine_cache)).count
```
Comment (ethn) [inline card/mod/admin/set/self/admin.rb]: Card::MachineCacheID?

File: card/mod/machines/lib/javascript/decko.js.coffee [hunk 3]
```diff
@@ -2,7 +2,7 @@ $.extend decko,
   # returns absolute path (starting with a slash)
   # if rawPath is relative (no slash), this adds relative root
```
Comment (ethn) [inline card/mod/machines/lib/javascript/decko.js.coffee]: need comment:
`# if rawPath is complete url, this returns the complete url`

File: card/mod/machines/set/abstract/machine.rb [hunk 4]
```diff
@@ -144,7 +144,7 @@ def direct_machine_input? input_card
 
 def run_engine input_card
   return unless direct_machine_input? input_card
-  if (cached = fetch_cache_card(input_card))
+  if (cached = fetch_cache_card(input_card)) && cached.content.present?
```
Comment (ethn) [inline card/mod/machines/set/abstract/machine.rb]: maybe we make
```
def content?
  content.present?
end
```
...that way we can do `&.content?` here and elsewhere.

File: card/mod/machines/set/all/reset_machines.rb [hunk 5]
```diff
@@ -1,12 +1,11 @@
...
   def reset_all_machines
     Auth.as_bot do
-      Card.search(
-        right: { codename: %w[in machine_cache machine_output] }
-      ).each do |card|
+      Card.search(right: { codename: "machine_output" }).each do |card|
         card.update_columns trash: true
         card.expire
       end
+      Card::Virtual.where(right_id: Card.fetch_id(:machine_cache)).delete_all
```
Comment (ethn) [inline card/mod/machines/set/all/reset_machines.rb]: Card::MachineCacheID again?

File: card/mod/machines/set/all/reset_machines.rb [hunk 6]
```diff
@@ -1,12 +1,11 @@
 module ClassMethods
   def reset_all_machines
     Auth.as_bot do
-      Card.search(
-        right: { codename: %w[in machine_cache machine_output] }
-      ).each do |card|
+      Card.search(right: { codename: "machine_output" }).each do |card|
```
Comment (ethn) [inline card/mod/machines/set/all/reset_machines.rb]: not that it matters, but I think search takes a block, so the "each" isn't necessary.

File: card/mod/virtual/lib/card/virtual.rb [hunk 7]
```diff
@@ -0,0 +1,77 @@
+# -*- encoding : utf-8 -*-
+
```
Comment (ethn) [inline card/mod/virtual/lib/card/virtual.rb]: I had a moment where I thought maybe this name should be CardVirtualCache or CardCacheVirtual or some variant with "content" in it.

My main concern was that users who didn't understand virtual cards would assume that to be virtual, you had to be in this table.

I'm not falling in love with those, so I think we can go with this, but we need some docs here explaining what this is (and maybe pointing to some not-yet-existent documentation of virtual cards).

File: card/mod/virtual/set/abstract/virtual.rb [hunk 8]
```diff
@@ -0,0 +1,45 @@
+# -*- encoding : utf-8 -*-
+
```
Comment (ethn) [inline card/mod/virtual/set/abstract/virtual.rb]: Can we make this name VirtualCache?

File: card/mod/machines/lib/javascript/decko_upload.js.coffee [hunk 9]
```diff
@@ -5,7 +5,7 @@ $.extend decko,
 # only the file, type_id and attachment_card_name are needed
 # attachment_card_name is the original card name,
 # ex: card[subcards][+logo][image], card[file]
-    $(fileupload).bind 'fileuploadsubmit', (e,data) ->
+    $(fileupload).on 'fileuploadsubmit', (e,data) ->
```
Comment (ethn) [inline card/mod/machines/lib/javascript/decko_upload.js.coffee]: wow, was that hard to find?

Comment (xithan) [reply card/mod/machines/lib/javascript/decko_upload.js.coffee]: not really, jquery has a changelog for that

File: card/mod/virtual/spec/set/abstract/virtual_cache_spec.rb [hunk 10]
```diff
@@ -0,0 +1,29 @@
+# -*- encoding : utf-8 -*-
+
+RSpec.describe Card::Set::Abstract::Virtual do
```
[the remark this answers is not in the document]
Comment (xithan) [reply card/mod/virtual/spec/set/abstract/virtual_cache_spec.rb]: me too

**Pull request merged**